File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.cc
#include <iostream>
#include <vector>
#include <queue>
#include <stack>
#include <algorithm>
using namespace std;

int N, M, V;
vector<vector<int>> graph;

void DFS() {
	vector<bool> visited;
	visited.resize(N + 1, false);

	stack<int> s;
	s.push(V);

	while (!s.empty()) {
		int curr = s.top();
		s.pop();

		if (visited[curr] == false) {
			visited[curr] = true;
			cout << curr << " ";

			sort(graph[curr].begin(), graph[curr].end());

			for (int i = graph[curr].size() - 1; i >= 0 ; i--)
			{
				int k = graph[curr][i];
				if (visited[k] == false) {
					s.push(k);
				}
			}
		}
	}
}

void BFS() {
	vector<bool> visited;
	visited.resize(N + 1, false);

	queue<int> q;
	q.push(V);

	while (!q.empty()) {
		int curr = q.front();
		q.pop();

		if (visited[curr] == false) {
			visited[curr] = true;
			cout << curr << " ";

			sort(graph[curr].begin(), graph[curr].end());

			for (int i = 0; i < graph[curr].size(); i++)
			{
				int k = graph[curr][i];
				if (visited[k] == false) {
					q.push(k);
				}
			}
		}
	}
}

int main() {
	cin >> N >> M >> V;
	graph.resize(N + 1);

	for (int i = 0; i < M; i++)
	{
		int u, v;
		cin >> u >> v;
		graph[u].push_back(v);
		graph[v].push_back(u);
	}

	DFS();
	cout << endl;
	BFS();
	return 0;
}

File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.py
import sys
input = sys.stdin.readline

[N, M, V] = list(map(int, input().strip().split(' ')))
graph = [[] for _ in range(N+1)]
for _ in range(M):
  [u, v] = list(map(int, input().strip().split(' ')))
  graph[u].append(v)
  graph[v].append(u)

def dfs(N, V, graph):
  stack = [V]
  visited = [False for _ in range(N+1)]
  result = []
  while(len(stack) > 0):
    curr = stack.pop()
    if not visited[curr]:
      visited[curr] = True
      result.append(curr)
    for node in reversed(sorted(graph[curr])):
      if not visited[node]:
        stack.append(node)
  
  result = list(map(str, result))
  return result

def bfs(N, V, graph):
  queue = [V]
  visited = [False for _ in range(N+1)]
  result = []

  while(len(queue) > 0):
    curr = queue.pop(0)
    if not visited[curr]:
      visited[curr] = True
      result.append(curr)
    for node in sorted(graph[curr]):
      if not visited[node]:
        queue.append(node)
  
  result = list(map(str, result))
  return result

print(' '.join(dfs(N,V,graph)))
print(' '.join(bfs(N,V,graph)))